prov/gni/src/gnix_msg.c

```diff
@@ -2474,22 +2494,43 @@ ssize_t _gnix_recv(struct gnix_fid_ep *ep, uint64_t buf, size_t len,
 			req->work_fn = req->msg.send_iov_cnt == 1 ?
 				__gnix_rndzv_req : __gnix_rndzv_iov_req_build;
 
-			ret = _gnix_vc_queue_work_req(req);
 
-			/* If using FI_MULTI_RECV and there is space left in
-			 * the receive buffer, try to match another unexpected
-			 * request. */
+			/*
+			 * Check if we're using FI_MULTI_RECV and there is
+			 * space left in the receive buffer.
+			 */
+
 			if ((req->msg.recv_flags & FI_MULTI_RECV) &&
 			    ((len - req->msg.cum_send_len) >= ep->min_multi_recv)) {
+				/* Allocate new request for this transfer. */
+				dup_req = __gnix_msg_dup_req(req);
+				if (!dup_req) {
+					COND_RELEASE(ep->requires_lock,
+						     queue_lock);
+					return -FI_ENOMEM;
+				}
```

Comment:
we need a separate request for every rendezvous match because we need to generate a distinct ```cqe->buf``` for each matched request.  The original code didn't do this and was one of the reasons evan had to comment out the ```cqe->buf``` check in the rdm_sr.c multirecv tests.